### PR TITLE
Refactoring trait pipeline (1/2)

### PR DESCRIPTION
**Pull request by linwang9926**
- Reorganizing the trait_map dictionary in the format: "trait_name": (function, ["input_trait1", "input_trait2", ...], {"additional_kwarg1": True, "additional_kwarg2": 0.5})
- modify ellipse, network functions by adding an Optional argument(s). Take ellipse for example:
  - def get_ellipse_a(pts_all_array: np.ndarray,ellipse: Optional[Tuple[float, float, float]],). 
  - The argument "ellipse" is ellipse feature, the values will be saved to the traits dictionary once we did trait calculation and BFS for the edges (traits).

==================================================================================
- Update ellipse and network functions argument. If the arguments have different types (Tuple vs. np.ndarray), then use a Union  to make this arguments inputting as pts_all_array or ellipse Tuple. If they are the same type, them make two arguments. for example:
  - get_ellipse_a(**pts_all_array: Union[np.ndarray, Tuple[float, float, float]]**)
  - get_network_solidity(primary_pts: np.ndarray,lateral_pts: np.ndarray,
      **chull_area: float = None,pts_all_array: Optional[np.ndarray] = None**,monocots: bool = False,)
- Modify the ellipse and network test functions
- Reorganize the arguments in trait_map dictionary in `get_traits_value_frame` function.

==================================================================================
- Update `get_root_angle` function by adding node_ind as an arugment.
- Update base-related functions (`get_base_ct_density`, `get_base_length_ratio`) by assuming the primary_pts as primary_length_max.
- Update `get_base_median_ratio` function by using lateral_base_ys and primary_tip_pt_y.
- Modify argument pts as Optional in `get_base_tip_dist` and `get_grav_index` functions
- Draft the trait_definitions using the defined TraitDef class
- Redo the `get_node_ind` function:
  - if proximal, get the nearest non-Nan node index.
  - If proximal, just use the first half root (if odd number node, the centre node is included).
  - If not proximal, just use the last half root (if odd number node, the centre node is included).
  - if there is no non-Nan node detected, just return an index of 0 (base node).

### CODE REVIEW

**Review by eberrigan — COMMENTED** [review]
looks good!

File: sleap_roots/bases.py [hunk 1]
```diff
@@ -24,15 +25,17 @@ def get_bases(pts: np.ndarray, monocots: bool = False) -> np.ndarray:
 
 
 def get_base_tip_dist(
```
Comment (eberrigan) [inline sleap_roots/bases.py:27]: the input traits always have to match the trait definition. This function is too complicated. We should simplify it to just assume taking in the expected input traits from from the trait graph.

File: sleap_roots/lengths.py [hunk 2]
```diff
@@ -107,7 +108,7 @@ def get_root_lengths_max(pts: np.ndarray) -> np.ndarray:
 def get_grav_index(
```
Comment (eberrigan) [inline sleap_roots/lengths.py:108]: Same here: only input traits in the trait definition can be arguments to the function.

File: sleap_roots/bases.py (hunk 1 above)
Comment (linwang9926) [reply sleap_roots/bases.py:27]: The `base_pts` and `tip_pts` can be used for our pipeline to quickly calculate the base-tip distance. The optional `pts` is designed for general purpose.

**Comment by linwang9926** [comment]
I redo the `get_node_ind` function:
- if proximal, get the nearest non-Nan node index.
- If proximal, just use the first half root (if odd number node, the centre node is included).
- If not proximal, just use the last half root (if odd number node, the centre node is included).
- if there is no non-Nan node detected, just return an index of 0 (base node).

**Comment by linwang9926** [comment]
Network-related functions () need further test after testing `get_root_lengths` function.

File: sleap_roots/angle.py [hunk 3]
```diff
@@ -41,18 +41,36 @@ def get_node_ind(pts: np.ndarray, proximal: bool = True) -> np.ndarray:
         return np.nan
 
     if proximal:
```
Comment (eberrigan) [inline sleap_roots/angle.py:43]: sorry I tried to vectorize this but looks like I did it incorrectly. If it was better before you can change it back!

Comment (linwang9926) [reply sleap_roots/angle.py:43]: No problem! The current version also works and uses vectors instead of a for loop.

**Pull request closed without merging**